File: Test_cos_se.m
function [avgError,flag]=Test_cos_se()
a=round(rand(1,1000)*360);%随机产生一个1*1000(double)数列，其中的值范围：0-360
%b=cosd(a);%以角度制计算cos函数，cosd是MATLAB固有函数
T=length(a);
%------------------------------------------------
%进行1000次的计算误差分析

  %将标准库中的cosd()与编写的cos_se()运算值进行误差计算：
     %对每一次误差计算的操作：
          %(1)两者误差值大于设定标准误差，即cos_se()计算错误，flag++
          %(2)两者误差值大于设定标准误差，即cos_se()计算无误
     %如上的操作重复1000（T）次，最终误差计算=测试计算中发生错误的次数/总测试数
for i=1:T %比较，计算错误的个数
    b(i)=cosd(a(i));%标准值
    c(i) = cos_se( a(i) );%测试值
    erro(i)=abs(b(i)-c(i));%计算每次误差=|标准值-测试值|
end
for j=1:T
if erro(j)>0.001
      flag=1;
      break;
  else
      flag=0;
end
 end
 avgError=sum(erro)/T;
end

File: Test_cos_se.py
#  -*-coding:utf8;-*-
#  qpy:3
#  qpy:console
import math
import random
from math import pi


def fa(a):
    b=1
    while a!=1:
        b*=a
        a-=1
    return b

def taylor(x,n):
    a=1
    count=1
    for k in range(1, n):
        if count%2!=0:
            a-=(x**(2*k))/fa(2*k)
        else:
            a+=(x**(2*k))/fa(2*k)
        count+=1
    return a

def cos_se(x):
    x = (x/180)*math.pi
    return taylor(x, 10)


def test_cos():
    num = 1000
    test_number = [0] * num
    sum = 0
    flag = 0  # 判断误差是否大于0.001的标志，默认为0，小于0.001
    for i in range(num):  # 随机选择1000个随机数
        test_number[i] = random.uniform(0, 360)
    for i in range(num):
        sta_val = math.cos(test_number[i] / 180 * pi)  # 标准值
        tes_val = cos_se(test_number[i])  # 需要测试的值
        error = abs(sta_val - tes_val)  # 误差
        if error > 0.001:
            flag = 1
            tes_val = math.cos(test_number[i] / 180 * pi)
            error = abs(sta_val - tes_val)
            if error <= 0.001:
                flag = 0
        sum = sum + error
    avgError = sum / num  # 平均误差
    return flag, avgError


if __name__ == '__main__':
    flag, avgError = test_cos()
    print('flag = ', flag)
    print('avgError = ', avgError)
